### PR TITLE
Add proper logging for stealing-type cards

### PR DESCRIPTION
**Pull request by erisalke**
Issue: https://github.com/terraforming-mars/terraforming-mars/issues/3175
Screenshot 2021-11-29 at 19 24 18
Screenshot 2021-11-29 at 19 24 27
Screenshot 2021-11-29 at 19 29 40

### CODE REVIEW

File: src/Player.ts [hunk 1]
```diff
@@ -348,6 +375,7 @@ export class Player implements ISerializable<SerializedPlayer> {
     options? : {
       log?: boolean,
       from? : Player | GlobalEventName,
+      isStealing?: boolean
```
Comment (kberg) [inline src/Player.ts:378]: I think shorten this to `stealing`

File: src/Player.ts [hunk 2]
```diff
@@ -317,12 +317,39 @@ export class Player implements ISerializable<SerializedPlayer> {
...
+      .player(this)
+    });
+  }
+
+  private logUnitDelta(
+    resource: Resources, 
+    amount: number, 
+    unitType: 'production' | 'amount', 
+    from: Player | GlobalEventName | undefined,
+    isStealing = false
```
Comment (kberg) [inline src/Player.ts:341]: short here to `stealing` too.

File: src/Player.ts [hunk 3]
```diff
@@ -317,12 +317,39 @@ export class Player implements ISerializable<SerializedPlayer> {
...
+  private logStealing(
+    resource: Resources, 
+    amount: number, 
+    unitType: 'production' | 'amount', 
+    from: Player
+  ) {
+    const absAmount = Math.abs(amount);
+    let message = '${0} stole ${1} ${2} ' + unitType + ' from ${3}';
+    this.game.log(message, (b) => {
+      b.player(from)
```
Comment (kberg) [inline src/Player.ts:329]: While I agree that your grammar is better than what logUnitDelta has, I expect the name order will bew confusing as players scan the log. This could be made simpler by just reusing logUnitDelta.

Change 
```
if (from !== undefined) {
  message = message + ' by ' + ((from instanceof Player) ? '${4}' : 'Global Event');
}
```

to

```
if (from !== undefined) {
  if (stealing === true) {
    message = message + ' stolen';
  }
  message = message + ' by ' + ((from instanceof Player) ? '${4}' : 'Global Event');
}
```

File: src/Player.ts [hunk 4]
```diff
@@ -409,7 +438,11 @@ export class Player implements ISerializable<SerializedPlayer> {
     }
   }
 
-  public addProduction(resource: Resources, amount : number, options? : { log: boolean, from? : Player | GlobalEventName}) {
+  public addProduction(
+    resource: Resources,
+    amount : number,
+    options? : { log: boolean, from? : Player | GlobalEventName, isStealing?: boolean}
```
Comment (kberg) [inline src/Player.ts:444]: Shorten `isStealing` here, too.

File: src/cards/base/AsteroidMiningConsortium.ts [hunk 5]
```diff
@@ -34,7 +34,9 @@ export class AsteroidMiningConsortium extends Card implements IProjectCard {
   }
 
   public play(player: Player) {
-    player.game.defer(new DecreaseAnyProduction(player, Resources.TITANIUM, 1));
+    const decreaseAction = new DecreaseAnyProduction(player, Resources.TITANIUM, 1);
```
Comment (kberg) [inline src/cards/base/AsteroidMiningConsortium.ts:37]: So if this were me I'd move much of the parameters to an object, just like `GainResources` The object would contain count (default to 1), title (defaulted to its current value) and stealing (default to false.) That'll be much simpler.

However, I'd probably do that in a completely separate PR or commit. If you like I'll do that.

**Review by kberg — CHANGES_REQUESTED** [review]
Thanks for thinking about this! Let me know if you want me to knock out generalizing DecreaseProduction etc.

File: src/Player.ts (hunk 3 above)
Comment (erisalke) [reply src/Player.ts]: Keep the same message format, sure thing.

File: src/cards/base/AsteroidMiningConsortium.ts (hunk 5 above)
Comment (erisalke) [reply src/cards/base/AsteroidMiningConsortium.ts]: I extracted the `Options` as follows:
```
export interface Options {
  count: number,
  stealing?: boolean
}
```
Therefore, all the previous calls changed 
from: `player.game.defer(new DecreaseAnyProduction(player, Resources.TITANIUM, 1));`
to: `player.game.defer(new DecreaseAnyProduction(player, Resources.TITANIUM, {count: 1}));`
and the stealing calls look this way:
`player.game.defer(new DecreaseAnyProduction(player, Resources.TITANIUM, {count: 1, stealing: true}));`

One thing I did't include into the `Options` was the title. For the moment all the existing calls use the default value - so it makes more sense to remove the parameter completely instead and hide it inside the class itself. And composing the "options" where the title is optional, but have a default value, and such value depends on the "count" which as well has a default value and may be optional... it's just messy.
I believe it's "good enough" for now. Maybe in future PRs if you really believe it's necessary.

File: src/cards/base/EnergyTapping.ts [hunk 6]
```diff
@@ -33,7 +33,8 @@ export class EnergyTapping extends Card implements IProjectCard {
 
   public play(player: Player) {
     player.addProduction(Resources.ENERGY, 1);
-    player.game.defer(new DecreaseAnyProduction(player, Resources.ENERGY, 1));
+    player.game.defer(
+      new DecreaseAnyProduction(player, Resources.ENERGY, {count:1, stealing:true}));
```
Comment (kberg) [inline src/cards/base/EnergyTapping.ts:37]: So one thing about reformatting on the fly is that it makes reviews more complicated.

image

Now I have to read the whole thing, whereas if you had just kept it on one line I could just the latter half of the line.

Don't make any changes, because I'll have read all of them by the time you get this.

File: src/cards/base/AsteroidMiningConsortium.ts [hunk 7]
```diff
@@ -34,7 +34,12 @@ export class AsteroidMiningConsortium extends Card implements IProjectCard {
   }
 
   public play(player: Player) {
-    player.game.defer(new DecreaseAnyProduction(player, Resources.TITANIUM, 1));
+    const decreaseAction = new DecreaseAnyProduction(
+      player,
+      Resources.TITANIUM,
+      {count: 1, stealing: true}
```
Comment (kberg) [inline src/cards/base/AsteroidMiningConsortium.ts:40]: Is this "stealing"? It looks like it, but is it?

File: src/deferredActions/DecreaseAnyProduction.ts [hunk 8]
```diff
@@ -3,24 +3,33 @@ import {Resources} from '../Resources';
...
-        public count: number = 1,
-        public title: string = 'Select player to decrease ' + resource + ' production by ' + count + ' step(s)',
-  ) {}
+    public player: Player,
+    public resource: Resources,
+    public options: DecreaseAnyProduction.Options = {
+      count: 1,
+      stealing: false
+    },
+    public title: string = 'Select player to decrease ' + resource + ' production by ' + options.count + ' step(s)',
```
Comment (kberg) [inline src/deferredActions/DecreaseAnyProduction.ts:21]: I agree with your assessment of title. I think I'd still put it in options, as options tends to be the last parameter. But don't bother, this is great.

File: src/deferredActions/DecreaseAnyProduction.ts [hunk 9]
```diff
@@ -41,7 +50,7 @@ export class DecreaseAnyProduction implements DeferredAction {
       this.title,
       'Decrease',
       (found: Player) => {
-        found.addProduction(this.resource, -this.count, {log: true, from: this.player});
+        found.addProduction(this.resource, -this.options.count, { log: true, from: this.player, stealing: this.options.stealing });
```
Comment (kberg) [inline src/deferredActions/DecreaseAnyProduction.ts:53]: There's no `decreaseProduction` method, is there? That's fine.

File: src/cards/base/PowerSupplyConsortium.ts [hunk 10]
```diff
@@ -34,7 +34,8 @@ export class PowerSupplyConsortium extends Card implements IProjectCard {
 
   public play(player: Player) {
     player.addProduction(Resources.ENERGY, 1);
-    player.game.defer(new DecreaseAnyProduction(player, Resources.ENERGY, 1));
+    player.game.defer(
+      new DecreaseAnyProduction(player, Resources.ENERGY, {count:1, stealing:true}));
```
Comment (kberg) [inline src/cards/base/PowerSupplyConsortium.ts:38]: Should there be a StealProduction deferred action? Not in this PR, just thinking.

**Review by kberg — COMMENTED** [review]
This is a good change. Normally I'd want to see some tests, but we kind of avoid many of the logging tests. If you can manually reverify these, this will be good enough with no more changes. (Or if you want to write a unit test, I won't mind! :D)

Once this is verified, I'll merge it in. Thank you!

Thanks!

File: src/cards/base/AsteroidMiningConsortium.ts (hunk 5 above)
Comment (kberg) [reply src/cards/base/AsteroidMiningConsortium.ts]: As you said, I also believe this is "good enough." You saw some of my comments lamenting that the API that existed before you came in here could have been better, but those can be dealt with another time. Thank you for this!

File: src/cards/base/AsteroidMiningConsortium.ts (hunk 7 above)
Comment (erisalke) [reply src/cards/base/AsteroidMiningConsortium.ts:40]: Depends how you define "stealing" I believe if one player lost "X" and other player gains exactly the same "X" (in this case titanium production) it is best described as "stealing" :D

File: src/deferredActions/DecreaseAnyProduction.ts (hunk 9 above)
Comment (erisalke) [reply src/deferredActions/DecreaseAnyProduction.ts:53]: nope, there's only the `addProduction` used with negative values

**Comment by kberg** [comment]
> This is a good change. Normally I'd want to see some tests, but we kind of avoid many of the logging tests. If you can manually reverify these, this will be good enough with no more changes. (Or if you want to write a unit test, I won't mind! :D)
> 
> Once this is verified, I'll merge it in. Thank you!
> 
> Thanks!

Thanks for the responses, etc. Just a reminder that my request above remains. Thanks again!

**Comment by erisalke** [comment]
@kberg All the changes were tested and works fine. You can see the results in the screens attached on the top. I think we could merge it now.

**Comment by kberg** [comment]
Those at the top are from before you made changes to the code, so it needs
another verification.

**Comment by erisalke** [comment]
@kberg ah no, sorry for not being clear enough 😓 
When I said: "All the changes were tested and works fine", I meant that I've **just finished** re-testing this.
In other words, it was checked that all works fine AFTER the last commit was made. It's javascript, we take no risks :)

**Review by kberg — APPROVED** [review]

**Comment by kberg** [comment]
AHA! Thank you. This looks good. Thank you!
